### DISCUSSION TITLE
Deploy to production on-demand, preserve PR previews

### DISCUSSION THREAD

**Discussion by sarahdayan** [vercel/vercel#6497] [Help]
Hi Vercel team!

I have a library organized as a monorepo, which contains the documentation site (built with Next.js, deployed on Vercel). This makes it easy for contributors to add documentation in a single PR, for reviewers to see a preview of the docs for each PR, and for the docs and library to share information easily and always be in sync.

I want to be able to hold off a release to npm until I have more stuff to release. For example, I might release at most once a week. This means I also want to hold off docs release. Conversely, I want to be able to push hotfixes to the docs (typos, broken links) without waiting. A solution is to go for a loose [Gitflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow): merge new features to a `next` branch, and do release snapshots on a `main` branch (which would trigger a release). I'd also configure Vercel to deploy on `main`.

There's still a lingering issue: **Vercel deploys on push by default.** This means that when I release, docs might arrive too early, documenting features that aren't there yet for an extra couple of minutes. For some docs features, I rely on the library being published (e.g., to inject the latest version on `window` on-demand from jsDelivr) so some features can be broken until the release is final. If the release fails for any reason, that's a lot of manual stuff to undo.

Ideally, **I'd prefer deploying on demand**, once the release is successful. This is theoretically doable with [Deploy Hooks](https://vercel.com/docs/more/deploy-hooks) and using the [Ignored Build Step](https://vercel.com/support/articles/how-do-i-use-the-ignored-build-step-field-on-vercel): I could write an ignore script that looks for something like `[docs release]` in the `VERCEL_GIT_COMMIT_MESSAGE` to allow a deploy. However, I'm wondering how well it plays out with PR previews. I don't want to have to ask contributors to add this in their commits to trigger a deploy. Maybe a solution would be to filter out on `VERCEL_GIT_COMMIT_REF`: if it isn’t `main`, I’d bypass the aforementioned check.

Remaining constraints would still be to find a way to add `[docs release]` in release commits (doable with automations) AND in hotfix commits (would have to be manual).

I’m thinking out loud here, mostly because even though it seems like a scenario that could work, it comes with tradeoffs and still feels like a complicated scenario where many things could go wrong. I’d love to know if I’m overthinking it and there’s a simpler way to do it before I start spending time on this. **Any opinions on such a flow?**

**Comment by leerob** [comment]
Hey! A couple of thoughts:

- We have started thinking about a solution allowing you to deploy blocking until certain CI actions have been completed, as part of the Vercel <> GitHub Integration.
- In the meantime, I think you could also write a custom GitHub Action and use the Vercel CLI to achieve this too.

**Comment by barelyhuman** [comment]
**Possible Solutions**
- Vercel CLI via Custom Github Actions 
- Deploy Hooks on Vercel that you can trigger on demand post the npm publish 
- The easiest one will be the separate branch concept but that will require more manual work in the long run 


**What I would go with**

0. Setup tokens on vercel  and small bash script to create .vercel/project.json 
1. Local NPM Publish 
2. Trigger a github release for the version change (using tags or whatever)
3. Github action that waits for creation of a release, post that run the vercel cli with the --token  option to exec a authorized build

**Things that go in the secrets**
1. GITHUB_TOKEN - if using a release action , this will be needed so that a release is created by you instead of the github-bot
2. ORG_ID , PROJECT_ID - needed for .vercel/project.json 
3. VERCEL_TOKEN - needed for the cli authentication of vercel
